plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -606,25 +609,81 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            DeletePrefixResult deletePrefixResult;
+            try {
+                deletePrefixResult = deletePrefix(path);
+            }
+            catch (AmazonClientException e) {
+                throw new IOException("Failed to delete paths with the prefix path " + path, e);
+            }
+            if (deletePrefixResult == DeletePrefixResult.NO_KEYS_FOUND) {
+                // If the provided key is not a "directory" prefix, attempt to delete the object with the specified key
+                deleteObject(key);
+            }
+            else if (deletePrefixResult == DeletePrefixResult.DELETE_KEYS_FAILURE) {
+                return false;
             }
+            deleteObject(key + DIRECTORY_SUFFIX);
         }
-        catch (FileNotFoundException e) {
-            return false;
+        else {
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(2), true);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, TrinoS3FileSystem::keysFromRecursiveListing));
+            if (objectKeysIterator.hasNext()) {
+                String childKey = objectKeysIterator.next();
+                if (!Objects.equals(childKey, key + PATH_SEPARATOR) || objectKeysIterator.hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleteObject(childKey);
+            }
+            else {
+                // Avoid deleting the bucket in case that the provided path points to the bucket root
+                if (!key.isEmpty()) {
```

Comment:
Can we move this check to an `key.isEmpty()` early return as the first thing in the not recursive block?

Reply:
Moving the check up would cause the method to return `false` while trying to delete empty keys even when the bucket is not empty - which would not hold the contract of the operation.
See `io.trino.plugin.hive.s3.TestTrinoS3FileSystemMinio#testDeleteNonRecursivelyNonEmptyBucketRoot`.